### PR TITLE
Add table `aws_networkfirewall_firewall`. Closes #1624

### LINKED ISSUE #1624: Add table `aws_networkfirewall_firewall`
Terraform link : https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/networkfirewall_firewall
SDK Link : https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/networkfirewall@v1.24.0#Client.ListFirewalls

### PR DESCRIPTION
**Pull request by madhushreeray30**
# Integration test logs
<details>
  <summary>Logs</summary>

```
Outputs:

account_id = "111122223333"
aws_region = "us-east-1"
resource_aka = "arn:aws:network-firewall:us-east-1:111122223333:firewall/turbottest75783"
resource_id = "arn:aws:network-firewall:us-east-1:111122223333:firewall/turbottest75783"
resource_name = "turbottest75783"
resource_policy_arn = "arn:aws:network-firewall:us-east-1:111122223333:firewall-policy/turbottest75783"
resource_tags = tomap({
  "name" = "turbottest75783"
})
resource_vpc_id = "vpc-058829c1efae32163"

Running SQL query: test-get-query.sql
[
  {
    "arn": "arn:aws:network-firewall:us-east-1:111122223333:firewall/turbottest75783",
    "name": "turbottest75783",
    "policy_arn": "arn:aws:network-firewall:us-east-1:111122223333:firewall-policy/turbottest75783",
    "region": "us-east-1",
    "tags": {
      "name": "turbottest75783"
    },
    "vpc_id": "vpc-058829c1efae32163"
  }
]
✔ PASSED

Running SQL query: test-get-tags-query.sql
[
  {
    "tags_src": [
      {
        "Key": "name",
        "Value": "turbottest75783"
      }
    ]
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "arn": "arn:aws:network-firewall:us-east-1:111122223333:firewall/turbottest75783",
    "name": "turbottest75783",
    "policy_arn": "arn:aws:network-firewall:us-east-1:111122223333:firewall-policy/turbottest75783",
    "region": "us-east-1",
    "tags": {
      "name": "turbottest75783"
    },
    "vpc_id": "vpc-058829c1efae32163"
  }
]
✔ PASSED

Running SQL query: test-notfound-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "account_id": "111122223333",
    "akas": [
      "arn:aws:network-firewall:us-east-1:111122223333:firewall/turbottest75783"
    ],
    "region": "us-east-1",
    "title": "turbottest75783"
  }
]
✔ PASSED

POSTTEST: tests/aws_networkfirewall_firewall

TEARDOWN: tests/aws_networkfirewall_firewall

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>

```
> select
  arn,
  name,
  description,
  vpc_id,
  policy_arn,
  region,
  tags
from
  aws_networkfirewall_firewall;
+---------------------------------------------------------------------------------+-----------------------+-----------------------+-----------------------+-----------------------------------------------------------------------------------------+------------+--------------------------------------------------+
| arn                                                                             | name                  | description           | vpc_id                | policy_arn                                                                              | region     | tags                                             |
+---------------------------------------------------------------------------------+-----------------------+-----------------------+-----------------------+-----------------------------------------------------------------------------------------+------------+--------------------------------------------------+
| arn:aws:network-firewall:ap-south-1:111122223333:firewall/test-nfirewall-130323 | test-nfirewall-130323 | test-nfirewall-130323 | vpc-0ca3f256c1c1b4680 | arn:aws:network-firewall:ap-south-1:111122223333:firewall-policy/test-nfirewallp-130323 | ap-south-1 | {"env":"test","subnet-change-protection":"true"} |
+---------------------------------------------------------------------------------+-----------------------+-----------------------+-----------------------+-----------------------------------------------------------------------------------------+------------+--------------------------------------------------+

> select
  f.arn,
  f.name,
  f.vpc_id,
  s ->> 'SubnetId' as subnet_id,
  cidr_block,
  availability_zone,
  default_for_az
from
  aws_networkfirewall_firewall f,
  jsonb_array_elements(subnet_mappings) s,
  aws_vpc_subnet vs
where
  vs.subnet_id = s ->> 'SubnetId';
+---------------------------------------------------------------------------------+-----------------------+-----------------------+--------------------------+----------------+-------------------+----------------+
| arn                                                                             | name                  | vpc_id                | subnet_id                | cidr_block     | availability_zone | default_for_az |
+---------------------------------------------------------------------------------+-----------------------+-----------------------+--------------------------+----------------+-------------------+----------------+
| arn:aws:network-firewall:ap-south-1:111122223333:firewall/test-nfirewall-130323 | test-nfirewall-130323 | vpc-0ca3f256c1c1b4680 | subnet-0c2ef4c65c858c2a4 | 10.20.0.0/26   | ap-south-1a       | false          |
| arn:aws:network-firewall:ap-south-1:111122223333:firewall/test-nfirewall-130323 | test-nfirewall-130323 | vpc-0ca3f256c1c1b4680 | subnet-0279e2ae238e316ed | 10.20.0.128/26 | ap-south-1c       | false          |
+---------------------------------------------------------------------------------+-----------------------+-----------------------+--------------------------+----------------+-------------------+----------------+
```
</details>

### CODE REVIEW

File: aws/table_aws_networkfirewall_firewall.go [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+			IgnoreConfig: &plugin.IgnoreConfig{
+				ShouldIgnoreErrorFunc: shouldIgnoreErrors([]string{"InvalidRequestException", "ValidationException"}),
+			},
+			Hydrate: listNetworkFirewallFirewalls,
+		},
+		GetMatrixItemFunc: SupportedRegionMatrix(networkfirewallv1.EndpointsID),
+		Columns: awsRegionalColumns([]*plugin.Column{
+			{
+				Name:        "arn",
+				Description: "The last time that the firewall policy was changed.",
```
Comment (ParthaI) [inline aws/table_aws_networkfirewall_firewall.go:41]: Please update the description.

File: aws/table_aws_networkfirewall_firewall.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+			{
+				Name:        "id",
+				Description: "The unique identifier for the firewall.",
+				Type:        proto.ColumnType_STRING,
+				Hydrate:     getNetworkFirewallFirewall,
+				Transform:   transform.FromField("Firewall.FirewallId"),
+			},
+			{
+				Name:        "policy_arn",
+				Description: "The public subnets that Network Firewall is using for the firewall. Each subnet must belong to a different Availability Zone.",
```
Comment (ParthaI) [inline aws/table_aws_networkfirewall_firewall.go:84]: Is the description correct?

File: aws/table_aws_networkfirewall_firewall.go [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+				Name:        "subnet_change_protection",
+				Description: "A setting indicating whether the firewall is protected against changes to the subnet associations.",
+				Type:        proto.ColumnType_BOOL,
+				Hydrate:     getNetworkFirewallFirewall,
+				Transform:   transform.FromField("Firewall.SubnetChangeProtection"),
+			},
+
+			{
+				Name:        "encryption_configuration",
+				Description: "A complex type that contains the Amazon Web Services KMS encryption configuration settings for your firewall.",
```
Comment (ParthaI) [inline aws/table_aws_networkfirewall_firewall.go:106]:
```suggestion
				Description: "A complex type that contains the Amazon Web Services KMS encryption configuration settings for the firewall.",
```

File: aws/table_aws_networkfirewall_firewall.go [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+			},
+		}),
+	}
+}
+
+//// LIST FUNCTION
+
+func listNetworkFirewallFirewalls(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
+
+	// vpcId := d.EqualsQuals["vpc_id"].GetStringValue()
```
Comment (ParthaI) [inline aws/table_aws_networkfirewall_firewall.go:162]:
```suggestion
```

File: aws/table_aws_networkfirewall_firewall.go [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+	}
+
+	if d.Quals["vpc_id"] != nil {
+		for _, q := range d.Quals["vpc_id"].Quals {
+			value := q.Value.GetStringValue()
+			if q.Operator == "=" {
+				input.VpcIds = append(input.VpcIds, value)
+			}
+		}
+	}
```
Comment (ParthaI) [inline aws/table_aws_networkfirewall_firewall.go:188]: Could we minimize the above code block as follows?

```
vpcId := d.EqualsQuals["vpc_id"].GetStringValue()
if strings.TrimSpace(vpcId) != "" {
		input.VpcIds = append(input.VpcIds, vpcId)
	}
```

File: aws/table_aws_networkfirewall_firewall.go [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+	// Reduce the basic request limit down if the user has only requested a small number of rows
+	maxLimit := int32(100)
+	if d.QueryContext.Limit != nil {
+		limit := int32(*d.QueryContext.Limit)
+		if limit < maxLimit {
+			if limit < 1 {
+				maxLimit = 1
+			} else {
+				maxLimit = limit
+			}
```
Comment (ParthaI) [inline aws/table_aws_networkfirewall_firewall.go:173]:
```suggestion
		maxLimit = limit
```

**Review by ParthaI — COMMENTED** [review]
@karanpopat Please see the comments. Thanks!

File: aws/table_aws_networkfirewall_firewall.go (hunk 5 above)
Comment (karanpopat) [reply aws/table_aws_networkfirewall_firewall.go:188]: This reduces api calls when the query has `in` clause. All in clause ids can be passed to the as the part of request params

**Review by misraved — APPROVED** [review]

**Pull request merged**